### PR TITLE
feat(ingest/kafka): support metadata mapping from kafka avro schemas

### PR DESCRIPTION
**Pull request by mayurinehate**
Adaptation of this PR: https://github.com/datahub-project/datahub/pull/7381 with slightly different implementation and more tests. Please refer the original PR for detailed description of feature.

In short, this PR allows dbt-style meta-mapping for dataset and field level additional properties present in avro schema of kafka topic. This also allows translating tags from existing tags listed in avro schema to DataHub tags. 

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by hsheth2** [comment]
@mayurinehate fyi there's a test failing on this one

I'll try to review this in the next two days

**Comment by mayurinehate** [comment]
> @mayurinehate fyi there's a test failing on this one
> 
> I'll try to review this in the next two days

I've fixed the test. Thank you for informing.

File: metadata-ingestion/src/datahub/ingestion/source/confluent_schema_registry.py [hunk 1]
```diff
@@ -59,6 +61,14 @@ def __init__(
         except Exception as e:
             logger.warning(f"Failed to get subjects from schema registry: {e}")
 
+        self.field_meta_processor = OperationProcessor(
+            self.source_config.field_meta_mapping,
+            self.source_config.tag_prefix,
+            "SOURCE_CONTROL",
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/confluent_schema_registry.py:67]: I'm not sure what this should be. "SOURCE_CONTROL" is probably okay for dbt. Should we set the ownership source to "OTHER" or "MANUAL"here ?

File: metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py [hunk 2]
```diff
@@ -251,12 +263,18 @@ def __init__(
...
+            tag_prefix: Optional[str] = None,
         ):
             self._schema = schema
             self._actual_schema = actual_schema
             self._converter = converter
             self._description = description
             self._default_value = default_value
+            self._meta_mapping_processor = meta_mapping_processor
+            self._schema_tags_field = schema_tags_field
+            self._tag_prefix = tag_prefix
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py:277]: we can get these using `converter._something` right? is there a reason to pass them explicitly?

File: metadata-ingestion/src/datahub/ingestion/source/confluent_schema_registry.py (hunk 1 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/confluent_schema_registry.py:67]: let's use `OwnershipSourceTypeClass.SERVICE`

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py [hunk 3]
```diff
@@ -89,6 +92,29 @@ class KafkaSourceConfig(StatefulIngestionConfigBase, DatasetSourceConfigMixin):
...
+        default={},
+        description="mapping rules that will be executed against field-level schema properties. Refer to the section below on meta automated mappings.",
+    )
+    strip_user_ids_from_email: bool = pydantic.Field(
+        default=False,
+        description="Whether or not to strip email id while adding owners using meta mappings.",
+    )
+    tag_prefix: str = pydantic.Field(
+        default="", description="Prefix added to tags during ingestion."
+    )
```
[the remark this answers is not in the document]
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/kafka.py:117]: honestly not sure why we added tag prefix at all

if we did want to support it, it should be implemented as a workunit helper right?

so anyways, empty is fine for now, with a goal of removing that field with something cleaner in the future

File: metadata-ingestion/src/datahub/utilities/mapping.py [hunk 4]
```diff
@@ -121,9 +125,18 @@ def process(self, raw_props: Dict[str, Any]) -> Dict[str, Any]:
                 )
                 if not operation_type or not operation_config:
                     continue
+                raw_props_value = raw_props.get(operation_key)
+                if not raw_props_value and self.match_nested_props:
+                    try:
+                        raw_props_value = reduce(
+                            operator.getitem, operation_key.split("."), raw_props
+                        )
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/utilities/mapping.py:133]: a comment would be helpful here

File: metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py (hunk 2 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/extractor/schema_util.py]: that makes sense. there is no reason to pass them explicitly.

File: metadata-ingestion/src/datahub/ingestion/source/confluent_schema_registry.py (hunk 1 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/confluent_schema_registry.py]: done

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/kafka.py:118]: I'm not sure either. Do you mean workunit processor ? Yes that would be better.

Also, we can remove prefix stuff from kafka it its not required.

**Review by hsheth2 — APPROVED** [review]

**Comment by hsheth2** [comment]
CI failure is unrelated

**Pull request merged**